viewing_party/party.py

```diff
@@ -1,23 +1,144 @@
+        watchlist.remove(movie_details)
+        watched.append(movie_details)
+    return updated_data
+        
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+    """Return the average rating of a user's watched movies. Return 0 if 
+    watched is empty."""
+    watched = user_data["watched"]
+    total = 0
+    for movie in watched:
+        total += movie["rating"]
+    return (total / len(watched)) if len(watched) else 0
+
+def get_most_watched_genre(user_data):
+    """Return the most watched genre in a user's watched list.""" 
+    genres = Counter([movie["genre"] for movie in user_data["watched"]])
+    return genres.most_common(1)[0][0] if genres else None
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def get_unique_watched(user_data):
+    """Return all the movies the user has watched that their friends have not 
+    watched."""
+    user_watched = {movie["title"]: movie for movie in user_data["watched"]}
+    friends_watched = set()
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_watched.add(movie["title"])
+    for title in friends_watched:
+        if title in user_watched:
+            user_watched.pop(title)
+    return list(user_watched.values())
+
+def get_friends_unique_watched(user_data):
+    """Return all the movies that at least one of the user's friends has 
+    watched but the user has not watched."""
+    friends_watched = {}
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_watched[movie["title"]] = movie
+    user_watched = {movie["title"] for movie in user_data["watched"]}
+    for title in user_watched:
+        if title in friends_watched:
+            friends_watched.pop(title)
+    return list(friends_watched.values())
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data):
+    """Return a list of movies that the user has not watched, that one of 
+    their friends has watched, and that is available on a service they are 
+    subscribed to."""
+    # Build list of movies that friends have watched.
+    friend_movies = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie not in friend_movies:
+                friend_movies.append(movie)
+    # Filter out movies not on a service the user is subscribed to.
+    recommended_movies = []
+    subscriptions = user_data["subscriptions"]
+    watched = user_data["watched"]
+    for movie in friend_movies:
+        if movie not in watched and movie["host"] in subscriptions:
+            recommended_movies.append(movie)
+    return recommended_movies

```

Comment:
👍  Nice work! Putting `user_data["subscriptions"]` and `user_data["watched"]` into variables definitely improved the readability of the combined conditional expression on line 110.

Reply:
How could we refactor this function to follow the single responsibility principle?